### PR TITLE
Zero Allocations

### PR DESCRIPTION
**Pull request by rf**
@raynos @jcorbin

Tries to have minimal interface changes by exposing methods that alloc the result objects, but the `pool`- prefix methods are available, which take a result object rather than allocing.

### CODE REVIEW

File: interface.js [hunk 1]
```diff
@@ -112,7 +112,8 @@ function genericResult(err, value, buffer, offset) {
         if (err.offset === undefined) err.offset = offset;
         if (err.buffer === undefined) err.buffer = buffer;
     }
-    return new Result(err, value);
+    var ret = new Result(err, value);
```
Comment (Raynos) [inline interface.js]: wat.

File: package.json [hunk 2]
```diff
@@ -18,7 +18,9 @@
   ],
   "dependencies": {
     "ansi-color": "^0.2.1",
+    "benchmark": "^2.0.0",
```
Comment (Raynos) [inline package.json]: wat.

File: package.json [hunk 3]
```diff
@@ -18,7 +18,9 @@
...
     "error": "^7.0.0",
+    "microtime": "^2.0.0",
```
Comment (Raynos) [inline package.json]: wat.

**Comment by Raynos** [comment]
Does what it says on tin, should get tests passing. Should link into tchannel & verify.

**Comment by kriskowal** [comment]
Perhaps instead of doubling the methods, we can introduce a final optional argument and default it to construct if absent.

```js
function f(blah, blah, blah, result) {
    result = result || new Result();
    if (err) {
        return result.reset(err);
    }
    return result;
}

**Comment by Raynos** [comment]
@kriskowal 

that makes it too easy to make mistakes.

Having two methods makes it easier because if you opt into pooling its mandatory and you know it will be pooled.

**Comment by kriskowal** [comment]
Having twice as much code is a greater liability. We should be migrating to the pooled-result model and drop the prior in the next major. Result object in callback position.

**Comment by rf** [comment]
We don't have twice as much code; the only methods I actually added are on the 3 on the BufferRW prototype, which allocate a result object and call out to the `pool`- prefix methods.

File: base.js [hunk 4]
```diff
@@ -81,34 +138,55 @@ function ReadResult(err, offset, value) {
...
+ReadResult.prototype.reset = function reset(err, offset, value) {
+    this.err = err;
+    this.offset = offset;
+    this.value = value;
+    return this;
+};
+
 ReadResult.error = function error(err, offset, value) {
-    return new ReadResult(err, offset, value);
+    throw new Error('deprecated');
```
Comment (kriskowal) [inline base.js]: This is backward-incompatible, not merely a deprecation. ThriftRW uses these API’s directly in some places, uses the constructor directly in others.

**Comment by kriskowal** [comment]
I see. I had assumed the the former public interface would be retained for backward compatibility.

I recommend a major version bump for this change, or a more graceful migration. If you would like a clean break, let’s make sure the new bufrw is 100% internally consistently using result args.

File: base.js (hunk 4 above)
Comment (rf) [inline base.js]: True, I can change this to something backwards compatible; the `throw`s were really there to assist the conversion (this is still a WIP).

**Comment by kriskowal** [comment]
Yeah, closer read again, let’s either consistently rename every method that now has a result argument, or not rename any methods at all. I also recommend putting the result argument in the final slot because it resembles a callback so closely. Nit: It can be called merely `result` over `destResult`.

**Comment by kriskowal** [comment]
Observation: not renaming any functions, putting result in the final argument slot, and defaulting it to a new result instance would make a graceful migration possible. The new interface would be a strict superset of the current. We could cut a major that warns if the result is not provided, and then cut another major that removes the defaulting code outright later. Added benefit is that the interface continues to feel internally consistent at every stage of the migration. We don’t want an endgame where every method has a pool* prefix, repeatedly stating the obvious.

**Comment by kriskowal** [comment]
Again, this is a *great* observation and I’m excited to see if this knocks ThriftRW into the neighborhood of JSON.

**Comment by rf** [comment]
I did consistently rename every method that now has a result argument :] and I put the arg first because some functions (like `byteLength`) take different numbers of params (some take a str or buf, like StringRW) so in order to have the compatible interface (i.e. [here](https://github.com/uber/bufrw/pull/57/files#diff-5656e7f7dd4df43f42db8d894c5396c0R61)) we'd have to know the number of params that function takes.

Other than the `throw`s, the interface is compatible and currently does not require a major version bump, though it does deliver performance improvements for free, since all aggregate RWs now pool internally.

File: base.js [hunk 5]
```diff
@@ -25,26 +28,65 @@ module.exports.LengthResult = LengthResult;
 module.exports.WriteResult = WriteResult;
 module.exports.ReadResult = ReadResult;
 
-function BufferRW(byteLength, readFrom, writeInto) {
+function BufferRW(poolByteLength, poolReadFrom, poolWriteInto) {
```
Comment (kriskowal) [inline base.js]: BufferRW is arguably a public interface, however luckily, ThriftRW and TChannel do not use it directly.

File: atoms.js [hunk 6]
```diff
@@ -37,25 +37,25 @@ function AtomRW(width, readAtomFrom, writeAtomInto) {
 }
 inherits(AtomRW, BufferRW);
 
-AtomRW.prototype.byteLength = function byteLength() {
-    return new LengthResult(null, this.width);
+AtomRW.prototype.poolByteLength = function byteLength(destResult) {
```
Comment (kriskowal) [inline atoms.js]: AtomRW.prototype.byteLength et al are public interfaces that have been removed. These *are* used by ThriftRW directly.

File: base.js (hunk 4 above)
Comment (kriskowal) [inline base.js]: :+1:

File: index.js [hunk 7]
```diff
@@ -77,7 +77,7 @@ var strn = StringRW(varint.unsigned, 'utf8');
 module.exports.str1 = str1;
 module.exports.str2 = str2;
 module.exports.strn = strn;
-module.exports.String = StringRW;
+module.exports.StringRW = StringRW;
```
Comment (kriskowal) [inline index.js]: Withdrawing String is backward-incompat. This is used directly in ThriftRW. Duplicate and deprecate if you must.

**Comment by kriskowal** [comment]
Please link against ThriftRW and TChannel to validate. Otherwise, please continue in this vein!

File: atoms.js (hunk 6 above)
Comment (rf) [inline atoms.js]: No, it inherits from BufferRW, which provides the compatibility layer. The `BufferRW#byteLength` method will allocate the result object and then call `poolByteLength`.

Comment (kriskowal) [inline atoms.js]: :+1:

File: base.js [hunk 8]
```diff
@@ -25,26 +28,65 @@ module.exports.LengthResult = LengthResult;
...
-    if (typeof readFrom === 'function') this.readFrom = readFrom;
-    if (typeof writeInto === 'function') this.writeInto = writeInto;
+
+    if (typeof poolByteLength === 'function') this.poolByteLength = poolByteLength;
+    if (typeof poolReadFrom === 'function') this.poolReadFrom = poolReadFrom;
+    if (typeof poolWriteInto === 'function') this.poolWriteInto = poolWriteInto;
+
+    assert(typeof this.poolByteLength === 'function', 'rw must implement `poolByteLength`');
+    assert(typeof this.poolReadFrom === 'function', 'rw must implement `poolReadFrom`');
+    assert(typeof this.poolWriteInto === 'function', 'rw must implement `poolWriteInto`');
```
Comment (kriskowal) [inline base.js]: The above can be simplified by asserting the arguments existence then assigning unconditionally.

**Comment by rf** [comment]
Verified with both thriftrw and tchannel-node; they pass tests with (almost) no code changes (tchannel was missing an inheritance). No longer a WIP.

**Comment by Raynos** [comment]
get coverage to green lol.

File: base.js [hunk 9]
```diff
@@ -25,24 +28,110 @@ module.exports.LengthResult = LengthResult;
 module.exports.WriteResult = WriteResult;
 module.exports.ReadResult = ReadResult;
 
-function BufferRW(byteLength, readFrom, writeInto) {
+function BufferRW(byteLength, readFrom, writeInto, isPooled) {
```
Comment (kriskowal) [inline base.js]: one neat trick to retain backward compatibility :+1:

File: test/stream/chunk_reader.js [hunk 10]
```diff
@@ -28,20 +28,20 @@ var ChunkReader = require('../../stream/chunk_reader');
...
 var readErrorRW = {
-    width: 1,
-    readFrom: function() {
-        return ReadResult.error(new Error('boom'));
-    }
+    poolReadFrom: function(destResult) {return destResult.reset(new Error('boom'), 0);},
+    width: 1
 };
 
+readErrorRW.__proto__ = BufferRW.prototype;
```
Comment (kriskowal) [inline test/stream/chunk_reader.js]: Consider making a constructor just for the singleton. __proto__ asignment seems too clever, but I trust you know whether v8 will care. Not too worried about the module constructor getting optimized.

**Comment by kriskowal** [comment]
:ok:

**Comment by jcorbin** [comment]
:ok: with me (however I've not read the diff in excruciating detail, as it looks like @Raynos and @kriskowal have done in depth review); I almost feel that this change warrants a major bump, but maybe a minor will do?

**Comment by Raynos** [comment]
lgtm.

**Comment by rf** [comment]
Going to go with a minor version bump since thriftrw and tchannel-node both pass tests with this linked in

**Pull request merged**